File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/StepsRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.StepsRecord
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toNumberDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.StepsRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [StepsRecord] object to a [StepsRecordDto].
 */
internal fun StepsRecord.toDto(): StepsRecordDto = StepsRecordDto(
    id = metadata.id,
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
    endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    count = count.toNumberDto(),
)

/**
 * Converts a [StepsRecordDto] to a Health Connect [StepsRecord] object.
 */
internal fun StepsRecordDto.toHealthConnect(): StepsRecord = StepsRecord(
    count = count.value.toLong(),
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    startZoneOffset = startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    endZoneOffset = endZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    metadata = metadata.toHealthConnect(id),
)

File: packages/health_connector_hk_ios/lib/src/mappers/health_record_mappers/steps_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show StepsRecord, HealthRecordId, sinceV1_0_0, Number;
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hk_ios/src/mappers/measurement_unit_mappers.dart';
import 'package:health_connector_hk_ios/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart'
    show StepsRecordDto, NumberDto;
import 'package:meta/meta.dart' show internal;

/// Converts [StepsRecord] to [StepsRecordDto].
@sinceV1_0_0
@internal
extension StepsRecordToDto on StepsRecord {
  StepsRecordDto toDto() {
    return StepsRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      zoneOffsetSeconds: startZoneOffsetSeconds,

      metadata: metadata.toDto(),
      count: count.toDto() as NumberDto,
    );
  }
}

/// Converts [StepsRecordDto] to [StepsRecord].
@sinceV1_0_0
@internal
extension StepsRecordDtoToDomain on StepsRecordDto {
  StepsRecord toDomain() {
    return StepsRecord(
      id: id?.toDomain() ?? HealthRecordId.none,
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime),
      startZoneOffsetSeconds: zoneOffsetSeconds,
      endZoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      count: count.toDomain() as Number,
    );
  }
}
